adrs/0049-prebuilt-rubies-at-runtime.md

```diff
@@ -3,43 +3,78 @@
+- Ideally works across all supported scenarios including self-hosted runners.
+- Ideally works on all the supported runner platforms (windows, mac-os, all linux variants)
+- Ideally works with actions container scenarios
 
-## Decisions
+## Options
 
-### Prebuilt Rubies
+### Download Prebuilt Rubies on Misses
 
 Other tools like offer a [distribution](https://nodejs.org/dist/) which offers a [queryable endpoint](https://nodejs.org/dist/index.json) which allows the desired version [by a version spec which is semver](https://github.com/actions/setup-node#setup-node).
 
 Other CI services like Travis support pulling a wide variety of [ruby versions offered here](http://rubies.travis-ci.org/).
 
-Requirements:
-- Offer prebuilt Ruby, JRuby, TruffleRuby versions 
-- Setup-Ruby exposes `ruby-version` which accepts a semver
-- Rubies are discoverable and queryable by the Setup-Ruby action.
-- Initially offer for all platforms offered by the hosted Actions VMs (Ubuntu 18, Mac 10.15 and Windows 2019)
-
 NOTE: The prebuilt rubies offered will only be supported for use by GitHub Actions.
 
 The scripts and tooling to build the Rubies will be at `actions/build-ruby` and it will be consumed by the existing action `actions/setup-ruby`.  Since the capabilities are additive, there's no need to create a `v2`.  There is no compat break.
 
-A specific version of a Ruby will be offered as an individual semi-immutable build-ruby repo release similar to how the [actions runner exposes versions](https://github.com/actions/runner/releases/tag/v2.164.0) where each version offers n platforms.  This is also similar to [Travis individual versions](http://rubies.travis-ci.org/ubuntu/18.04/s390x/ruby-2.6.5).
+A specific version of a Ruby will be offered as an individual build-ruby repo release similar to how the [actions runner exposes versions](https://github.com/actions/runner/releases/tag/v2.164.0) where each version offers n platforms.  This is also similar to [Travis individual versions](http://rubies.travis-ci.org/ubuntu/18.04/s390x/ruby-2.6.5).
+
+**Pros**
+
+On a toolkit cache miss, the first build is slower but not that slow since it's pre-built.
+
+**Cons**
+
+Will not work with self-hosted scenarios
+Will not work across all platforms the runner supports
+Will not work across various container scenarios
+Only supported on hosted VMs 
+
+### Build and Cache Misses at Runtime
+
+[rbenv/ruby-build](https://github.com/rbenv/ruby-build) offers a convenient way to build at runtime.
+
+Build on cache miss of an exact version.  Offer an option to cache using the upcoming [toolkit cache lib](https://github.com/actions/toolkit).  There are plans to factor out the caching logic in the [actions/cache](https://github.com/actions/cache) so other actions can take advantage of caching services.
+
+**Pros**
+
+Aligns well with the only supported Ruby distribution ... source code.
+Works on self-hosted, 
+
+**Cons**
+
+The first build on a miss will be slower but will not fail.
```

Comment:
That strategy would be similar to https://github.com/clupprich/ruby-build-action.
Yes, that will be *very* slow, like 5 minutes, and requires installing extra system dependencies at runtime: https://github.com/eregon/ruby-install-builder/runs/380224237

Reply:
It would also do so many duplicate builds of the same version in many repositories, i.e., I think it would waste resources significantly.

Reply:
See comment below.  We need to look at first build per org and only on cache miss ( folks that bound to specific version ).  The mainline minor binding bypasses all of this with vm cache of latest versions

Reply:
After the miss, it would be cached in the same data center for future builds and very fast downloads.  On self-hosted, it would live past that job and be in the machines cache and instantly use it.

Reply:
remember that the hosted VMs are caching the latest (and possibly n -1 ) of every minor version so this is only first build for folks that pin back to an old specific version.